### PR TITLE
Various GLES fixes found when using Mesa.

### PR DESCRIPTION
**Pull request by michaelrgb**
-Disable sRGB for GLES.
-glFramebufferTexture emulated with glFramebufferTexture2D
-glBindFragDataLocation not allowed for gl_ prefix outputs.

### CODE REVIEW

File: renderdoc/driver/gl/gl_hooks_egl.cpp [hunk 1]
```diff
@@ -435,12 +435,7 @@ __attribute__((visibility("default"))) EGLContext eglCreateContext(EGLDisplay di
   init.depthBits = value;
   eglhooks.eglGetConfigAttrib_real(display, config, EGL_STENCIL_SIZE, &value);
   init.stencilBits = value;
-// TODO: how to detect this?
-#if ENABLED(RDOC_ANDROID)
   init.isSRGB = 0;
```
Comment (michaelrgb) [inline renderdoc/driver/gl/gl_hooks_egl.cpp]: sRGB extensions for ES:
-EGL_KHR_gl_colorspace is only for default framebuffer created by EGL
-GL_EXT_sRGB is not complete
Therefore disable it for ES, whether its Android or desktop.

Comment (baldurk) [inline renderdoc/driver/gl/gl_hooks_egl.cpp]: `EGL_KHR_gl_colorspace` sounds like it's a match for what this code is intended for - for sRGB support for GL-created textures we can capture that normally with the format, this `isSRGB` flag is to mark if the default framebuffer is sRGB. That way when we create a fake stand-in texture we know the format to create.

I'm fine with this as a default but maybe a comment should be left marking a future improvement to detect use of `EGL_KHR_gl_colorspace` and set this bool?

File: renderdoc/driver/gl/gl_renderstate.cpp [hunk 2]
```diff
@@ -637,6 +637,7 @@ bool GLRenderState::CheckEnableDisableParam(GLenum pname)
       case eGL_PROGRAM_POINT_SIZE:
       case eGL_PRIMITIVE_RESTART:
       case eGL_TEXTURE_CUBE_MAP_SEAMLESS:
+      case eGL_FRAMEBUFFER_SRGB:
```
Comment (baldurk) [inline renderdoc/driver/gl/gl_renderstate.cpp]: I know there's some weird behaviour for GLES around SRGB where it varies subtly from GL. I don't remember the details but this came up a while ago when I was talking about GLES.

From what I can see GL_FRAMEBUFFER_SRGB isn't supported on GLES but instead of acting like it's always disabled (which is what these unsupported enabled bits end up being) it acts like it's always _enabled_. What is the impact of leaving the enabled bit false? Is there any fixup needed elsewhere in the codebase to handle the different meaning for GLES/GL, or is it better to have it just to silently default to enabled on GLES?

File: renderdoc/driver/gl/gl_renderstate.cpp [hunk 3]
```diff
@@ -1205,7 +1206,7 @@ void GLRenderState::ApplyState(void *ctx, WrappedOpenGL *gl)
   if(HasExt[ARB_shader_image_load_store])
   {
     GLuint maxImages = 0;
-    m_Real->glGetIntegerv(eGL_MAX_IMAGE_UNITS, (GLint *)&maxImages);
+    m_Real->glGetIntegerv(eGL_MAX_TEXTURE_IMAGE_UNITS, (GLint *)&maxImages);
```
Comment (baldurk) [inline renderdoc/driver/gl/gl_renderstate.cpp]: What's the reason for this change? `GL_MAX_TEXTURE_IMAGE_UNITS` is (confusingly) the number of active units you can select with `glActiveTexture`. It seems to be true for ES3 and desktop GL.

`GL_MAX_IMAGE_UNITS` is the number of image load/store bindings, which I think is what we want here.

Comment (michaelrgb) [inline renderdoc/driver/gl/gl_renderstate.cpp]: Ah sorry this is actually available. Yes I mixed them up.

File: renderdoc/driver/gl/gl_hooks_egl.cpp (hunk 1 above)
Comment (michaelrgb) [inline renderdoc/driver/gl/gl_hooks_egl.cpp]: IIUC, the problem with checking for EGL_KHR_gl_colorspace is that it is only for default framebuffers. 

I just tested Mesa and it does actually support EGL_KHR_gl_colorspace, so I pushed the EGL_GL_COLORSPACE_KHR=EGL_GL_COLORSPACE_SRGB_KHR attrib pair. This worked fine so I set isSRGB to true, but when I replayed the capture of course it creates the m_FakeBB_FBO **non-default** framebuffer. This reported the error:
'GL_INVALID_FRAMEBUFFER_OPERATION in glClear(incomplete framebuffer)'

So whilst I can create sRGB default EGL framebuffers, the replay code still wont work because that extension still wont let us FBOs with sRGB textures.

So would you agree that leaving a comment to check for EGL_KHR_gl_colorspace would be a bit misleading?

Comment (baldurk) [inline renderdoc/driver/gl/gl_hooks_egl.cpp]: Oh wow, I didn't realise that GLES doesn't allow creating FBOs with sRGB textures. I understand the problem now, I assumed that was easily possible.

That seems like a massive missing feature - does that mean that anyone wanting to do gamma correct rendering on GLES has to do manual shader sRGB conversion to any offscreen targets?

It might be worth still leaving a comment explaining this as why we are unable to properly replay a capture using EGL_KHR_gl_colorspace (since although the user is able to create an sRGB backbuffer, we'll be forced to create a non-corrected FBO as you point out). I see what you mean though, it is tricky.

Comment (michaelrgb) [inline renderdoc/driver/gl/gl_hooks_egl.cpp]: Yes I will leave exactly that comment.

File: renderdoc/driver/gl/gl_renderstate.cpp (hunk 2 above)
Comment (michaelrgb) [inline renderdoc/driver/gl/gl_renderstate.cpp]: I see what you mean, you were referring to this in the EGL_KHR_gl_colorspace spec:
"GL ES doesn't require GL_FRAMEBUFFER_SRGB enable."

Because we can never create sRGB framebuffers (due to above issue), we have to assume that its always disabled, and therefore I think this new case label is correct.

File: renderdoc/driver/gl/gl_hooks_egl.cpp (hunk 1 above)
Comment (michaelrgb) [inline renderdoc/driver/gl/gl_hooks_egl.cpp]: I just had another read of the ES 3.2 spec pdf, GL_SRGB8 is marked as not color-renderable (table 8.10), which is why it gives incomplete framebuffers. 

Ive added that to the comment. Bizarrely SRGB8_ALPHA8 is color-renderable, so I suppose we could check if EGL_BUFFER_SIZE was 32, and then set isSRGB=1.

Comment (baldurk) [inline renderdoc/driver/gl/gl_hooks_egl.cpp]: Interesting, that makes more sense then. I'm not surprised a 24-bit texture isn't color-renderable, since I guess hardware will mostly want 32-bit pixel strides. Not important to fix right now but it's good to know for the future that this is soluble.

Comment (michaelrgb) [inline renderdoc/driver/gl/gl_hooks_egl.cpp]: Ill try fixing it now: check for the EGL attrib as well as 32bit buffer size. Hold off on the commit.

File: renderdoc/driver/gl/gl_hooks_egl.cpp [hunk 4]
```diff
@@ -509,6 +505,12 @@ __attribute__((visibility("default"))) EGLBoolean eglSwapBuffers(EGLDisplay dpy,
   eglhooks.eglQuerySurface_real(dpy, surface, EGL_HEIGHT, &height);
   eglhooks.eglQuerySurface_real(dpy, surface, EGL_WIDTH, &width);
 
+  GLInitParams &init = eglhooks.GetDriver()->GetInitParams();
+  int colorspace;
+  eglhooks.eglQuerySurface_real(dpy, surface, EGL_COLORSPACE, &colorspace);
+  // GL_SRGB8_ALPHA8 is specified as color-renderable, unlike GL_SRGB8.
+  init.isSRGB = init.colorBits == 32 && colorspace == EGL_COLORSPACE_sRGB;
```
Comment (michaelrgb) [inline renderdoc/driver/gl/gl_hooks_egl.cpp]: I think this is the correct check to determine isSRGB.

Not ideal to make GLInitParams public - I just realised I can probably do this calculation inside eglMakeCurrent.

Comment (galpeter) [inline renderdoc/driver/gl/gl_hooks_egl.cpp]: based on the egl.h/docs the `EGL_COLORSPACE` and `EGL_COLORSPACE_sRGB` constant are related to OpenVG and not to GL ES. We would need to use `EGL_GL_COLORSPACE` and `EGL_GL_COLORSPACE_sRGB` but that is only available from EGL 1.5 or via the EGL_KHR_gl_colorspace extension.

Comment (michaelrgb) [inline renderdoc/driver/gl/gl_hooks_egl.cpp]: Right, but GL has copied these values for interchangeability.

If EGL_KHR_gl_colorspace is not supported, then the query would leave the return value unmodified on failure. Are you satisfied if I initialize the colorspace value to 0 so we will do the safe thing of setting isSRGB=0 ?

Comment (galpeter) [inline renderdoc/driver/gl/gl_hooks_egl.cpp]: I think the constants are different (from the egl.h ver 1.5):
```
#define EGL_COLORSPACE                    0x3087
#define EGL_COLORSPACE_sRGB               0x3089
```
```
#define EGL_VG_COLORSPACE                 0x3087
#define EGL_VG_COLORSPACE_sRGB            0x3089
```
```
#define EGL_GL_COLORSPACE                 0x309D
#define EGL_GL_COLORSPACE_SRGB            0x3089
```
from the EGL_KHR_gl_colorspace extension:
```
EGL_GL_COLORSPACE_KHR                   0x309D
EGL_GL_COLORSPACE_SRGB_KHR              0x3089
```

That aside: having the colorpsace as a defult 0 is good imho.

Comment (michaelrgb) [inline renderdoc/driver/gl/gl_hooks_egl.cpp]: Switched to the constants from 1.5/EGL_KHR_gl_colorspace and initialized the returned value.

Thanks for pointing this out, the previous enums had no affect on my test app framebuffer's gamma, these do.
